src/libOpenImageIO-C/c-imageio_defines.h

```diff
@@ -0,0 +1,37 @@
+// Copyright Contributors to the OpenImageIO project.
```

Comment:
c-imageio_defines.h: We can always straighten this out later, but I wonder if this should just go in c-imageio.h rather than being a separate file that has no direct analogy in the C++ API? Was there a reason to break it into a separate header?

Reply:
At the time I did it it was to separate definitions into things that would be visible to C and to C++ to avoid redefining symbols, but that doesn't seem necessary anymore so yeah I'll collapse this into c-imageio.h

Reply:
Ahh no that definitely was still necessary. Basically we can't include the c-imageio.h in c-imageio.cpp because the C interface forward declares opaque ImageSpec, ImageInput etc., and then puns them in c-imageio.cpp by importing OIIO::ImageInput etc into the TU namespace. If we include the opaque definitions then we get lots of compiler errors because I'm essentially defining ImageInput twice. 

I'm halfway through doing it "properly" and removing the pointer punning and it's turning into a bit of a casting nightmare:
```
ImageInput*
ImageInput_open(const char* filename, const ImageSpec* config, void* _ioproxy)
{
    auto ii = OIIO::ImageInput::open(
        filename, reinterpret_cast<const OIIO::ImageSpec*>(config), nullptr);
    return reinterpret_cast<ImageInput*>(ii.release());
}
```

I think a better solution here might be to return to punning but change the C type names to prepend `OIIO_`, which to be honest we should probably do anyway in the name of hygeine. So the above would then become:
```
using OIIO_ImageInput = OIIO::ImageInput;
using OIIO_ImageSpec = OIIO::ImageSpec;

OIIO_ImageInput*
ImageInput_open(const char* filename, const OIIO_ImageSpec* config, void* _ioproxy)
{
    return OIIO::ImageInput::open(filename, config, nullptr).release();
}
```
That then begs the question of should we be prepending `OIIO_` to all the function names as well, which again my gut tells me yes, even though it's getting a little verbose.

Reply:
Yeah, that probably the right thing to do. 

It's easy to hate on C++, and I can complain plenty, but it's been interesting seeing this process and being reminded of all the parts that I would be very frustrated by missing if I were working in C again. Namespaces and function overloading by argument is definitely something I really like. I really wish the C++ language itself dictated the name mangling, so that you could easily construct a C linkage entry point name from a C++ function. That would dramatically simplify the amount of new code required.

Reply:
Namespaces I’m with you on for sure. Not so convinced about function overloading any more. Rust doesn’t allow it either and despite some initial frustration I’ve actually come to really like it. I find it much easier to read as you don’t have to mentally type-check every function call to know what it’s doing.